Neo4j migrations

ProGM opened the pull request:
This pull introduces an `ActiveRecord` like migration method.

It's W.I.P., but I have a small thing I want to discuss before proceeding:

I'm using this conventions: migration files are searched in `db/neo4j/migrate` folder using the same strategy as ActiveRecord.
Migrations state is saved in a `SchemaMigration` node.

Since i wanted to copy the ActiveRecord implementation, I created this set of tasks:

```bash
rake neo4j::migrate # Alias of ::all
rake neo4j::migrate::all
rake neo4j::migrate::up VERSION=some_version
rake neo4j::migrate::down VERSION=some_version
rake neo4j::rollback
```

However, the first one is conflicting the current neo4j migration task. For now, I renamed it `rake neo4j::legacy_migrate`. What do you think about it?

Pings:
@cheerfulstoic
@subvertallchris

cheerfulstoic commented:
I would probably rename it `migrate_task`, though @subvertallchris  wrote the migration stuff initially.  What do you think Chris?

subvertallchris commented:
We can change the old task but I think that `migrate` and `migrate_task` is confusing. Can we think of a better name?

I like the way this is implemented but I'm not sure how it'll work in practice since all of our schema settings are mixed throughout models. What would your ideal workflow be to take advantage of this?

ProGM commented:
An alternative naming convention could be keeping the old `neo4j::migrate`, and renaming the new migrations like this:

```bash
rake db:neo4j:migrate # Alias of ::all
rake db:neo4j:migrate::all
rake db:neo4j:migrate::up VERSION=some_version
rake db:neo4j:migrate::down VERSION=some_version
rake db:neo4j:rollback
```

>  What would your ideal workflow be to take advantage of this?

I think that neo4j migrations will be used mostly with cypher queries.

Some examples:

"Drop a table"-like migration
```ruby
class DropNodeMyLabel < Neo4j::Migrations::Base
  def up
    execute 'MATCH (n:`MyLabel`)-[r]-() DELETE n, r'
  end
  def down
    raise Neo4j::IrreversibleMigration
  end
end
```

Renaming a property.
```ruby
class RenameTitleToName < Neo4j::Migrations::Base
  def up
    execute 'MATCH (c:Category) where c.name is null SET c.name = c.title REMOVE c.title'
  end
  def down
    execute 'MATCH (c:Category) where c.title is null SET c.title = c.name REMOVE c.name'
  end
end
```

Maybe we could define some helpers for this actions, like `rename_property` or `drop_nodes`, or something similar :)

What do you think?

@subvertallchris @cheerfulstoic

subvertallchris commented:
Ah, I see. For tasks like that, yes, it'd be really nice to have and makes sense as written. I was fixated on the index and constraint definitions.

As for naming, we can deprecate them entirely if you want. We could move the UUID task somewhere else entirely. I don't want to confuse people with similar names and this makes more sense as _the_ `migrate`.

cheerfulstoic commented:
Yeah, I don't think there would be many schema migrations, though we'd probably want to have `drop_index` and `drop_constraint` helpers since removing the model definitions don't remove the them.

I think that `rename_property` would be good.  Probably it would take a label?

What would `drop_nodes` do?  Would that be like `Model.delete_all`?

Thinking through other cases, perhaps `add_label`, though I could see that as being something that we add to scoping in general like `Model.where(foo: 1).add_label(:Bar)`.  Though I guess that could just be `Model.where(foo: 1).query_as(:m).set('m:Bar').exec`.

ProGM commented:
@cheerfulstoic @subvertallchris I did it, please check 👍

(Any better idea to test migration is welcome 😆)

cheerfulstoic commented on lib/neo4j.rb:
```diff
@@ -86,6 +86,13 @@
+require 'neo4j/migrations/base'
+require 'neo4j/migrations/runner'
+require 'neo4j/migrations/schema_migration'
+
```
I feel like this `require` practice is growing a bit too much ;)  I wonder about using Ruby's `autoload`:

http://www.rubyinside.com/ruby-techniques-revealed-autoload-1652.html

Migrations will generally run in the rake task process and then the process dies.  So if we use `autoload` we (I think) won't get the `Migration` code loaded into memory in the application, just in the rake task.

If it works well here, I think it's something that we see if it makes sense elsewhere.

ProGM replied:
I'll take a look 👍

cheerfulstoic commented on lib/neo4j/migrations/helpers.rb:
```diff
@@ -0,0 +1,50 @@
+  module Migrations
+    module Helpers
+      def rename_property(label, old_property, new_property)
+        by_label(label).where(n: {new_property => nil})
```
I don't know that I would expect that this would only work on nodes where the `new_property` is `nil`.

My first instinct was to just blow away any values which are on the `new_property`, but perhaps we should run a check to see if the `new_property` is non-null anywhere and give an error if that's the case.

ProGM replied:
Adding a test for this ;)

ProGM replied:
@cheerfulstoic Maybe I could fail by default if there's some new_property `nil`, and add a parameter `force`?

cheerfulstoic replied:
I'm down with that.  @subvertallchris ?

cheerfulstoic commented on lib/neo4j/migrations/helpers.rb:
```diff
@@ -0,0 +1,50 @@
+
+      private
+
+      delegate :query, to: Neo4j::Session
```
This `query` would be nice as a public helper.

ProGM replied:
👍

cheerfulstoic commented on lib/neo4j/migrations/helpers.rb:
```diff
@@ -0,0 +1,50 @@
+      end
+
+      def drop_nodes(label)
+        query.match("(n:`#{label}`)")
```
You can also do: `query.match(n: label)`, I think.

Kudos on not going with the `DETACH DELETE` (though of course there's no method for it.  Just put a note on my list to add one ;)

ProGM replied:
first done.

About DETACH DELETE, what should I do?

cheerfulstoic replied:
We can't use `DETACH DELETE` as we still support Neo4j 2.1 / 2.2.

ProGM replied:
Uh, I see 👍

cheerfulstoic commented on lib/neo4j/migrations/helpers.rb:
```diff
@@ -0,0 +1,50 @@
+      end
+
+      def add_labels(label, *labels)
+        by_label(label).set("n:#{labels.join(':')}").exec
```
You might be able to do `.set(n: labels)` here.  If not it's something we might add

ProGM replied:
done 👍

cheerfulstoic commented on lib/neo4j/migrations/helpers.rb:
```diff
@@ -0,0 +1,50 @@
+      alias remove_label remove_labels
+
+      def remove_constraint(label, property)
+        execute("DROP CONSTRAINT ON (n:`#{label}`) ASSERT n.#{property} IS UNIQUE")
```
I believe that you can do something like `Neo4j::Label.new(label). create_constraint(property, type: :unique)`.  @subvertallchris ?

subvertallchris replied:
Indeed you can. See https://github.com/neo4jrb/neo4j-core/blob/master/lib/neo4j-server/cypher_label.rb and https://github.com/neo4jrb/neo4j-core/blob/master/lib/neo4j/label.rb. The interface might still have some rough edges. I gave it an overhaul a few months ago but it was borne from a very Java perspective that wasn't exactly friendly...

ProGM replied:
Done 👍

cheerfulstoic commented on lib/neo4j/migrations/migration_file.rb:
```diff
@@ -0,0 +1,24 @@
+      end
+
+      def create
+        require file_name
```
Not a huge deal, but I generally prefer to refer to instance variables internally (like `@file_name`), not the accessors

ProGM replied:
done 👍

cheerfulstoic commented on lib/neo4j/migrations/migration_file.rb:
```diff
@@ -0,0 +1,24 @@
+
+      def extract_data!
+        @version, @symbol_name = File.basename(file_name, '.rb').split('_', 2)
+        @class_name = @symbol_name.split('_').map(&:capitalize).join('')
```
Would `@symbol_name.classify` work?

ProGM replied:
Is ActiveSupport included? If so, ok 👍

cheerfulstoic replied:
Yup!  We use the modules `PerThreadRegistry`, `Notifications`, `Concern`, and a few core extensions, so it's pretty ingrained into the gem by now ;)  But we try to require only the sections that we need

cheerfulstoic replied:
Yup!  We use the modules `PerThreadRegistry`, `Notifications`, `Inflector`, `Concern`, and a few core extensions, so it's pretty ingrained into the gem by now ;)  But we try to require only the sections that we need

cheerfulstoic replied:
Yup!  We use the modules `PerThreadRegistry`, `Notifications`, `Inflector`, `Concern`, and a few core extensions, so it's pretty ingrained into the gem by now ;)  But we try to require only the sections that we need

ProGM replied:
Triple comment, good work github :metal:

cheerfulstoic commented on lib/neo4j/migrations/runner.rb:
```diff
@@ -0,0 +1,106 @@
+      FILE_MISSING = '**** file missing ****'.freeze
+      STATUS_TABLE_HEADER = ['Status'.freeze, 'Migration ID'.freeze, 'Migration Name'.freeze].freeze
+      UP_MESSAGE = 'up'.freeze
+      DOWN_MESSAGE = 'down'.freeze
```
![http://s3-ec.buzzfed.com/static/2014-01/enhanced/webdr02/7/18/anigif_enhanced-buzz-13598-1389137896-5.gif](http://s3-ec.buzzfed.com/static/2014-01/enhanced/webdr02/7/18/anigif_enhanced-buzz-13598-1389137896-5.gif)

ProGM replied:
haha 😆 😆

cheerfulstoic commented on lib/neo4j/migrations/runner.rb:
```diff
@@ -0,0 +1,106 @@
+      DOWN_MESSAGE = 'down'.freeze
+
+      def initialize
+        @up_versions = SchemaMigration.all.pluck(:migration_id)
```
Probably a minor performance improvement, but if this were a `Set` then the `include?` calls would be faster.

ProGM replied:
👍

cheerfulstoic commented on lib/neo4j/migrations/runner.rb:
```diff
@@ -0,0 +1,106 @@
+
+      def all
+        migration_files.each do |migration_file|
+          next if @up_versions.include?(migration_file.version)
```
Since this is done so much, maybe a private method called something like `up?`?

ProGM replied:
(done)

cheerfulstoic commented on lib/neo4j/migrations/runner.rb:
```diff
@@ -0,0 +1,106 @@
+          migration_name = migration_file ? migration_file.class_name : FILE_MISSING
+          output STATUS_TABLE_FORMAT, status, version, migration_name
+        end
+      end
```
:+1: for this

cheerfulstoic replied:
:+1: for this method

cheerfulstoic commented on lib/neo4j/migrations/runner.rb:
```diff
@@ -0,0 +1,106 @@
+          migration = migration_file.create
+          migration.migrate(:down)
+        end
+      end
```
Maybe these methods could be one `migrate` method that takes a second `direction` argument of `:up` or `:down` and you could switch passing in `migrated` / `reverted` based on that argument.

ProGM replied:
done (refactored)

cheerfulstoic commented on lib/neo4j/migrations/runner.rb:
```diff
@@ -0,0 +1,106 @@
+      end
+
+      def output(*string_format)
+        puts format(*string_format) unless ENV['silenced']
```
I normally expect ENV vars to be all caps.  Also maybe it could be more specific like `MIGRATIONS_SILENCED`?

ProGM replied:
I copy-pasted it from the old `Migration` module:
https://github.com/neo4jrb/neo4j/blob/master/lib/neo4j/migration.rb#L10
However, it obviously makes sense

cheerfulstoic replied:
Ah, I didn't know that.  Meh, new migration, new major version, new behavior ;)  I would just make sure that it goes into the `CHANGELOG` under `Changes`

ProGM replied:
done!

cheerfulstoic commented on lib/neo4j/migrations/schema_migration.rb:
```diff
@@ -0,0 +1,11 @@
+    class SchemaMigration
+      include Neo4j::ActiveNode
+
+      property :migration_id, type: String, constraint: :unique
```
Perhaps this could be the `id_property` because it's a natural identifier and then we don't need to worry about `uuid` (which would just mean more data returned from the DB anyway)

cheerfulstoic commented on lib/neo4j/tasks/migration.rake:
```diff
@@ -21,4 +21,44 @@ namespace :neo4j do
+
+  desc 'A shortcut for neo4j::migrate::all'
+  task :migrate do
+    Rake::Task['neo4j::migrate::all'].invoke
```
I thought you only did a single `:` between namespaces in rake.  Can you do both?

ProGM replied:
No, it's a typo 👎 fixed!

cheerfulstoic commented on spec/e2e/migrations_spec.rb:
```diff
@@ -0,0 +1,113 @@
+describe Neo4j::Migrations::Runner do
```
One trick that I've seen is that you can surround your `RSpec` with the module(s) in question to reduce the amount of repetition of those modules.  For example here you could do:

```ruby
module Neo4j
  module Migrations
    describe Runner do
      # etc...
      before do
        # etc...
        SchemaMigration.delete_all
      end
      # etc...
    end
  end
end
```

ProGM replied:
Done!

cheerfulstoic commented on spec/e2e/migrations_spec.rb:
```diff
@@ -0,0 +1,113 @@
+describe Neo4j::Migrations::Runner do
+  before(:all) { ENV['silenced'] = 'true' }
+  after(:all)  { ENV['silenced'] = nil }
```
It's not a big deal for this I think, but we might think about having a `let_env_variable` helper which would work something like `let_env_variable('silenced', true)` and it would store and restore the variable back to it's original state after the spec is done (similar to `let_config`)

ProGM replied:
done!

cheerfulstoic commented on spec/unit/migrations/helpers_spec.rb:
```diff
@@ -0,0 +1,68 @@
+  describe '#rename_property' do
+    it 'renames a property' do
+      rename_property :Book, :name, :title
+      expect(Book.all(:n).pluck('n.title')).to include('Book1', 'Book2', 'Book3')
```
We should test whatever edge-case behavior we decide on regarding my comment above about what happens to nodes which already have the new property set.

cheerfulstoic commented on spec/unit/migrations/helpers_spec.rb:
```diff
@@ -0,0 +1,68 @@
+
+  describe '#add_labels' do
+    it 'adds labels to a node' do
+      add_labels :Book, :Item, :Readable
```
I think that there should actually be a second argument which is an `Array`.  Looking at this at first I thought you were adding three labels, and then I remembered the method definition.  Like so:

```ruby
add_labels :Book, [:Item, :Readable]
add_label :Book, :Item
```

ProGM replied:
Done!

cheerfulstoic commented on spec/unit/migrations/helpers_spec.rb:
```diff
@@ -0,0 +1,68 @@
+
+  describe '#remove_index' do
+    it 'removes an index from a property' do
+      remove_index :Book, :author_name
```
This could probably use the `Label` class from code to get the list of indexes and you could see if the `author_name` index exists

ProGM replied:
done!

cheerfulstoic commented:
I had a number of comments, but it looks really great overall!  Really excited to have this coming up, thanks!

A couple of things to make sure we address:

 * `CHANGELOG` should lay out anything that changed / was added
 * We should have docs.  I don't think that migrations are in readthedocs yet, but this seems like a good opportunity.  Probably this should be it's own section.

ProGM commented:
@cheerfulstoic @subvertallchris Ok, some updates: I fixed all issues you wrote, doing some refactoring.

I also added the `remove_property` helper and updated the CHANGELOG.

Please take a look, later I'll take care of docs 😭

subvertallchris commented:
Very cool

ProGM commented:
Ach, jruby broken 😞

cheerfulstoic commented:
This all looks amazing!  Thanks again!

Once docs are there and jruby is working, feel free to merge away!

subvertallchris commented:
...and when there's a test for the transactional stuff we talked about.

ProGM commented:
@cheerfulstoic @subvertallchris I did a bit of refactor to improve error messages.

Moreover I (hopefully) fixed jruby and added some tests for transactions.

Only docs are missing now. However, if you have time, take a look to this last commit, I did a few changes that may require a revision. 😜

cheerfulstoic commented:
I believe that `Transaction.run` takes a boolean argument to say if you want to have a transaction (strange, I know, but it's pretty much for this sort of thing).  

Also, completely optional, but I just wanted to make sure that this thought wasn't just it my head: we might also implement [`say`](http://apidock.com/rails/v3.0.9/ActiveRecord/Migration/say/class) and [`say_with_time`](http://apidock.com/rails/ActiveRecord/Migration/say_with_time)

Thanks!

ProGM commented:
Ok, I also added `say` and `say_with_time` (they are more or less copy-pasted from ActiveRecord, since it works as-is)

ProGM commented:
@cheerfulstoic @subvertallchris I added also `rename_label` helper and I created a `Migrations` section in the docs.

What do you think? :) (please check the grammar 😆 )

cheerfulstoic commented on docs/Migrations.rst:
```diff
@@ -0,0 +1,260 @@
+Migrations
+==================
+
+Neo4j.rb provides an ``ActiveRecord``-like migration framework and a set of helper methods to manipulate both database schema and data.
```
For people that are new to Neo4j they may need some context since the use of our migrations isn't quite like `ActiveRecord`'s.  Maybe saying something like "Neo4j does not have a set schema like relational databases, but sometimes changes to the schema and the data are required.  To help with this, Neo4j.rb provides an ``ActiveRecord``-like" etc...

ProGM replied:
👍

cheerfulstoic commented on docs/Migrations.rst:
```diff
@@ -0,0 +1,260 @@
+    ...
+  end
+
+Neo4j::Migrations::Helpers
```
Perhaps a title like "Migration Helpers" so that people don't think they need to know about the `Neo4j::Migrations::Helpers" class.

Also, perhaps the `RenameUserNameToFirstName` migration example should use at least one helper so that people can make the association about where they would call the methods below.

ProGM replied:
Sounds good, fixed

cheerfulstoic commented on docs/Migrations.rst:
```diff
@@ -0,0 +1,260 @@
+
+.. code-block:: ruby
+
+  query.match(:n).where(name: 'John').delete(:n)
```
To actually get this to execute they would need to end this with `exec` or some `Enumerable` method (`to_a`/`each`/`select`/etc...)

ProGM replied:
👍 fixed

cheerfulstoic commented on docs/Migrations.rst:
```diff
@@ -0,0 +1,260 @@
+:Cypher:
+  .. code-block:: cypher
+
+    MATCH (n:`User`) REMOVE n.money
```
In thinking about this I went back and forth, but I think we don't need the `Cypher`.  If we ever change how the helper executes then we'd need to remember to update the docs.

This makes me realize, though, that we **should** make sure we're displaying the `Cypher` which is being executed during a migration run.  Do you know if that's currently happening?  If not, it should be possible to just set the log level (see the `railtie.rb`)

ProGM replied:
Uhm, right. I have to check for output.

cheerfulstoic commented on docs/Migrations.rst:
```diff
@@ -0,0 +1,260 @@
+
+.. code-block:: ruby
+
+  drop_constraint(:User, :name)
```
Will this fail if you don't define `disable_transactions!` if you're only making schema changes?

ProGM replied:
Currently... yes. And maybe it's not a great idea, though.
Check:
https://github.com/neo4jrb/neo4j/pull/1197/files#diff-ab636d4950487b612dfbe9fde69e387bR48

It's better to leave the neo4j error pass through

cheerfulstoic commented:
Feedbacked!

ProGM commented:
@cheerfulstoic @subvertallchris travis is 👍 , what do you think?

To merge or not to merge? 💀

subvertallchris commented:
This is really fantastic, go for it!

cheerfulstoic commented:
![http://www.ninefeettall.com/wp-content/uploads/2014/10/Merge-sign-yellow.jpg](http://www.ninefeettall.com/wp-content/uploads/2014/10/Merge-sign-yellow.jpg) ❗

ProGM commented:
👍
